ci: Add model evaluation GitHub action

GrayNekoBean opened the pull request:
Add a new CI workflow that can be triggered manually to execute a full evaluation cycle. The workflow takes care of deploying the application, generating synthetic data, running the RAG Chat evaluation and red-teaming scripts, dumping the results as artifacts on GitHub, and purging the deployment at the end if explicitly specified.

To test this workflow before merging, you can trigger it manually using the GitHub CLI and specifying the development branch (i.e. `ci-config`) as well as any input parameters needed (please make sure you have installed the [GitHub CLI tool](https://cli.github.com/)):

```bash
gh workflow run .github/workflows/model-evaluation.yml \
  --repo ucl-contoso-chat/ucl-openai-search \
  --ref ci-config \
  -f azd-env-name=ucl-oai-search-ci \
  -f skip-deploy=false
  -f purge=false
```

This CI accepts 3 input parameters:
- `azd-env-name`: The name of the Azure environment used for the deployment (by default, the value of the `AZURE_ENV_NAME` repository variable will be used, i.e. `ucl-oai-search`). If an existing environment is provided here, the corresponding deployment will be modified.
- `skip-deploy`: Whether to skip the provisioning and deployment steps (default: false)
- `purge`: Whether to purge the deployment after the evaluation has finished (default: false)

GrayNekoBean committed:
feat: added model evaluation workflow for github action

- The model-evaluation workflow only accepts manual trigger, it will run a complete cycle of deploy, run evaluation and undeploy.

GrayNekoBean committed:
feat: added model evaluation workflow for github action

- The model-evaluation workflow only accepts manual trigger, it will run a complete cycle of deploy, run evaluation and undeploy.

phoevos commented:
@GrayNekoBean, rebased on top of `main` (it was pointing to a random branch, `hf-client-integration`).

phoevos reviewed (changes requested):
Thanks!

To begin with, we need instructions for testing this. The workflow relies on an evaluation script that's not merged yet. Where can one find this script? What do they need to do to use it? How can they trigger the workflow? What sort of information do they absolutely need to provide? I see that the action reads loads of things from the GH environment, are these already set?

Apart from that, I see that the workflow duplicates all code found in `azure-dev.yaml`. Is there any reason why we can't reuse that?

GrayNekoBean commented:
I'll be here explain one by one.

> To begin with, we need instructions for testing this. The workflow relies on an evaluation script that's not merged yet. Where can one find this script? 

The scripts are on the evaluation branch, it will need the whole directory of `./evaluation/`. But even we have those scripts, it's hard to test them locally, even with [this tool](https://github.com/nektos/act), you will need to download a 75G large image to test this workflow locally. That's also the reason I cannot test it locally after I introduced a certain step.

> What do they need to do to use it? How can they trigger the workflow? 

Here's a command to manually trigger this CI workflow but using the **original ci branch** (safely):
```
gh workflow run full-evaluation.yml --repo ucl-contoso-chat/ucl-openai-search --ref ci -f provision=false -f deploy=false -f purge=false 
```
I'll add this to my pr message if you think this is okay.

> What sort of information do they absolutely need to provide? I see that the action reads loads of things from the GH environment, are these already set?

There're some arguments to set whether to deploy or undeploy the online services. They all have a default value and I could add more instructions to them in the PR message. They can be specified using the command line arguments. For the GH environments, yes they are already properly set in this repository and works globally regardless of the branch. No more actions required to set environment variables.

> Apart from that, I see that the workflow duplicates all code found in `azure-dev.yaml`. Is there any reason why we can't reuse that?

Yes, in fact many fractions of code are from the `azure-dev.yaml`, earlier I've also thought about invoking that workflow file via call_dispatch but as the arguments were introduced and they separately control two steps for "provision" and "deploy", this way would not work properly anymore.

phoevos commented:
I don't get why we allow controlling the `deploy` and `provision` steps individually. Provisioning without deploying makes no sense. Deploying without provisioning is a valid scenario, when we wanna use existing infrastructure but update the application code for some reason. But if the infrastructure exists, the provisioning step will be a noop anyway. Even if these flags made sense, you would add them to `azure-dev` and have one workflow trigger the other instead of copy-pasting 100 lines of code.

phoevos commented:
> you will need to download a 75G large image to test this workflow locally

This sounds terribly alarming. I'm guessing it refers to the evaluation side so we'll discuss it there when the time comes.

GrayNekoBean commented:
> > you will need to download a 75G large image to test this workflow locally
> 
> This sounds terribly alarming. I'm guessing it refers to the evaluation side so we'll discuss it there when the time comes.

Sorry if this sentence sounds misguidance. That's the tool's problem, nothing related to our codebase. It's only because that tool needs to download a larger docker image to run the github action workflow locally, before it only needs a smaller one, but in this workflow it runs an "az login" step which requires its largest image.

GrayNekoBean commented:
> I don't get why we allow controlling the `deploy` and `provision` steps individually. Provisioning without deploying makes no sense. Deploying without provisioning is a valid scenario, when we wanna use existing infrastructure but update the application code for some reason. But if the infrastructure exists, the provisioning step will be a noop anyway. Even if these flags made sense, you would add them to `azure-dev` and have one workflow trigger the other instead of copy-pasting 100 lines of code.

okay I'll consider changing the mechanism of these two arguments and using workflow trigger

GrayNekoBean committed:
model-evaluation now reuse azure-dev.yml workflow for deploying

GrayNekoBean committed:
Merge branch 'ci-config' of github.com:ucl-contoso-chat/ucl-openai-search into ci-config

GrayNekoBean committed:
set the default value for arg "deploy-only" to false.

phoevos commented on .github/workflows/model-evaluation.yml:
```diff
@@ -0,0 +1,201 @@
+  contents: read
+
+jobs:
+    deploy-azure:
+      if: ${{ inputs.provision-and-deploy == true || inputs.deploy-only == true }}
+      uses: ./.github/workflows/azure-dev.yml
+      secrets: inherit
+      with:
+        deploy-only: ${{ inputs.deploy-only }}
```
This step deploys something based on the `azure-dev.yaml` workflow, which is responsible for setting up some resources on Azure. The immediate next step, however, has no idea about what was just deployed. Instead, it retrieves some values from the GitHub/repo environment, which I understand you've set in advance. This can't work! We need to be retrieving the actual environment variables set after `azure-dev` was done deploying. **That** is the deployment we want, not one that we happened to have configured at some arbitrary point in time.

GrayNekoBean replied:
I will look for a method to automate this, I might implement this inside the evaluation scripts or in a separate python script file, for each different variables there might be different way to retrieve their value. 

However, please allow me to clarify here, using our current configuration and the way we do the deployment each time, I can confirm the values are always identical regardless how many times it's redeployed (or provisioning). This is not "configured at some arbitrary point in time", you can think this as an initial configuration that needs to be done.

phoevos replied:
> I will look for a method to automate this, I might implement this inside the evaluation scripts or in a separate python script file, for each different variables there might be different way to retrieve their value.

No, there already is a way. Isn't a `.env` file created with all of the environment variables? That should be all we need. You need to ensure that this file is carried over to the next job or that its content is sourced into it.

> However, please allow me to clarify here, using our current configuration and the way we do the deployment each time, I can confirm the values are always identical regardless how many times it's redeployed (or provisioning).

Yes, they are, because we generate all names based on the name of the resource group (or less often the subscription). But what if something changes in the code? Why would someone have to update repo configurations following a change? They wouldn't as things like these create inconsistencies. We set environment variables and secrets on GH for things that are actually **constant**. And these are just the few values in our environment that have to be there when we first deploy: the environment name, the subscription ID, the location for all resources, the SKU and location for Search, and the location of Document Intelligence. There's no point in hardcoding everything else, they'll be populated once the provisioning succeeds or once the tool detects that the desired resources are already there.

> This is not "configured at some arbitrary point in time", you can think this as an initial configuration that needs to be done.

This is wrong because you don't know these names before deploying a specific version of the code. You don't know how these hashes included in the names are generated. In fact, without looking at the code you can't even know for sure that they're hashes. What if we change the way we name resources in the code? Even better, what if Bicep changes their hashing algorithm? You can't assume that all of these values will stay the same. If you wanted this to make sense, you could have selected manually picked names for all of these resources. But you really don't have to, because we really do want to just run the provisioning script every time and propagate all of the envvars it created to the next jobs/steps.

GrayNekoBean replied:
okay, I will use variables in the `.env` file to set the default values for eval script configs.

phoevos commented on .github/workflows/model-evaluation.yml:
```diff
@@ -0,0 +1,201 @@
+on:
+  workflow_dispatch:
+    inputs:
+      provision-and-deploy:
+        description: 'Whether to provision the deployment'
+        required: true
+        default: true
+        type: boolean
+      deploy-only:
+        description: 'Whether to deploy the application'
+        required: true
+        default: false
+        type: boolean
```
As mentioned before, these 2 flags make little sense: https://github.com/ucl-contoso-chat/ucl-openai-search/pull/11#issuecomment-2230716330

GrayNekoBean replied:
I already changed the logic here, cases like "run provision but no deploy" will not happen. If I change it to an enumeration config with 3 options like this:  `pre-processing: [nothing, deploy-only, provision-deploy]`, would that be better?

phoevos replied:
Did you read my comment?

>  But if the infrastructure exists, the provisioning step will be a noop anyway.

If the Azure Resource Manager detects that what you're trying to provision already exists without changes, it will skip provisioning. There's no need for logic like this here.

GrayNekoBean replied:
ok I see, I'll change that it to a single boolean option here.

phoevos replied:
The only option we need is the one controlling resource deletion!

phoevos commented on .github/workflows/azure-dev.yml:
```diff
@@ -1,6 +1,14 @@
 on:
+  workflow_call:
+    inputs:
+      deploy-only:
+        description: 'only run azd deploy with no provisioning'
+        required: true
+        default: false
+        type: boolean
```
Once again, due to the way ARM works, this doesn't make much sense: https://github.com/ucl-contoso-chat/ucl-openai-search/pull/11#issuecomment-2230716330

GrayNekoBean replied:
same above

phoevos commented on .github/workflows/model-evaluation.yml:
```diff
@@ -0,0 +1,201 @@
+              OPENAI_HOST: ${{ vars.OPENAI_HOST }}
+              OPENAI_GPT_MODEL: ${{ vars.OPENAI_GPT_MODEL }}
+              AZURE_OPENAI_EVAL_DEPLOYMENT: ${{ vars.AZURE_OPENAI_EVAL_DEPLOYMENT }}
+              AZURE_OPENAI_SERVICE: ${{ vars.AZURE_OPENAI_SERVICE }}
+              AZURE_SEARCH_SERVICE: ${{ vars.AZURE_SEARCH_SERVICE }}
```
Many of those variables have been set at the job level, there's no point in repeating them here. Please check for duplicates and avoid repetitions!

GrayNekoBean replied:
Got it, I'll clean up these duplicates.

phoevos commented on .github/workflows/model-evaluation.yml:
```diff
@@ -0,0 +1,201 @@
+          - name: Run AI Rag Evaluations
+            working-directory: ./evaluation
+            run: |
+              OPENAI_HOST=$OPENAI_HOST OPENAI_GPT_MODEL=$OPENAI_GPT_MODEL AZURE_OPENAI_EVAL_DEPLOYMENT=$AZURE_OPENAI_EVAL_DEPLOYMENT AZURE_OPENAI_SERVICE=$AZURE_OPENAI_SERVICE AZURE_SEARCH_SERVICE=$AZURE_SEARCH_SERVICE AZURE_SEARCH_INDEX=$AZURE_SEARCH_INDEX AZURE_PRINCIPAL_ID=$AZURE_PRINCIPAL_ID python -m scripts evaluate --config=./config.json --numquestions=10
```
Perhaps there's a more elegant way of setting shell/environment variables!

GrayNekoBean replied:
Yes there should be, I'll look it up.

phoevos commented on .github/workflows/model-evaluation.yml:
```diff
@@ -0,0 +1,201 @@
+            working-directory: ./evaluation
+            run: |
+              python -m pip install --upgrade pip
+              pip install -r scripts/requirements.txt
+              pip install pyrit
```
As we've established, this should have been raised in time and we should never have resorted to measures like that. We'll fix the requirements in our project so that we can avoid clashes.

GrayNekoBean replied:
I will remove the second line, just make sure that the pip conflict issue is resolved in the evaluation branch.

phoevos replied:
Absolutely, I'm dealing with it!

phoevos commented on .github/workflows/model-evaluation.yml:
```diff
@@ -0,0 +1,201 @@
+        runs-on: ubuntu-latest
+        strategy:
+          matrix:
+            os: ["ubuntu-20.04"]
```
Why not a newer one? I see above that the job runs on "ubuntu-latest".

GrayNekoBean replied:
my bad, I will remove line 40, that's accidentally introduced and in fact that version value isn't affecting anything.

phoevos commented on .github/workflows/model-evaluation.yml:
```diff
@@ -0,0 +1,201 @@
+        default: false
+        type: boolean
+
+permissions:
+  id-token: write
+  contents: read
```
Why do we need these for this workflow?

GrayNekoBean replied:
Because I'm using [Azure Login](https://github.com/Azure/login) workflow to login to azure. And this workflow requires these two permissions.

phoevos replied:
Got it, cheers!

phoevos commented on .github/workflows/model-evaluation.yml:
```diff
@@ -0,0 +1,201 @@
+jobs:
+    deploy-azure:
+      if: ${{ inputs.provision-and-deploy == true || inputs.deploy-only == true }}
+      uses: ./.github/workflows/azure-dev.yml
```
When you're not executing a script I don't think there's a reason to start your path with `./`
```suggestion
      uses: .github/workflows/azure-dev.yml
```

GrayNekoBean replied:
sure

phoevos commented on .github/workflows/model-evaluation.yml:
```diff
@@ -0,0 +1,201 @@
+              python-version: ${{ matrix.python_version }}
+              architecture: x64
+    
+          - name: Install azd
+            uses: Azure/setup-azd@v1.0.0
+    
+          - name: Install Nodejs
+            uses: actions/setup-node@v4
+            with:
+              node-version: 18
+
+          #--------------------------Deploy--------------------------------
+          - name: Check credentials
+            run: |
+              echo "AZURE_CLIENT_ID: ${{ env.AZURE_CLIENT_ID }}"
+              echo "AZURE_TENANT_ID: ${{ env.AZURE_TENANT_ID }}"
+              echo "AZURE_SUBSCRIPTION_ID: ${{ env.AZURE_SUBSCRIPTION_ID }}"
+              echo "AZURE_PRINCIPAL_ID: ${{ secrets.AZURE_PRINCIPAL_ID }}"
+
+          - name: Azure login
+            uses: azure/login@v2
+            with:
+              client-id: ${{ env.AZURE_CLIENT_ID }}
+              tenant-id: ${{ env.AZURE_TENANT_ID }}
+              subscription-id: ${{ env.AZURE_SUBSCRIPTION_ID }}
+            
+          - name: Azure CLI script
+            uses: azure/cli@v2
+            with:
+              azcliversion: latest
+              inlineScript: |
+                az account show
```
Are these actually needed as part of this job? Not for the evaluation part at least. Even if some of these are actually needed for azd down to run, that should be a separate job anyway.

GrayNekoBean replied:
Step "Check credentials" and "Azure CLI script" can be removed since that was meant for debugging purposes. But other steps are necessary, they are not for `azd down`, they are for running the evaluation scripts. Evaluation scripts need to invoke the OpenAI service that deployed on Azure, and it is using the credential information from [App Registration](https://portal.azure.com/#view/Microsoft_AAD_RegisteredApps/ApplicationMenuBlade/~/Credentials/appId/f53943c3-513e-4d52-8897-fe1dad4f8052/isMSAApp~/false) to login to azure to gain access to the Cognitive Service.

phoevos reviewed (changes requested):
@GrayNekoBean there are some issues with the current approach, please take a look at the comments!

GrayNekoBean committed:
Update .github/workflows/model-evaluation.yml

Co-authored-by: Phoevos Kalemkeris <phoevos@protonmail.com>

GrayNekoBean committed:
Update CI config

- Removed deployment-control input arguments
- Removed redundant and debugging codes
- Set default value for environment variables config using the azd deployment generated .env file

GrayNekoBean commented:
There will be more changes to our codebase in the foreseeable future, so constantly changing the CI workflow to keep it in sync with other branches is not an efficient exercise at this time. Additionally, I've noticed that frequent redeployments by CI could be interruptive to other team members' work. Therefore, work around the CI branch will be halt until the other branches are confirmed as complete.

phoevos committed:
Merge branch 'main' into ci-config

phoevos committed:
fix: Retrieve envvars from azd environment

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
feat: Make azd env configurable on workflow call

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

xiatianrui1110 committed:
fix: Add dataset generation step

Signed-off-by: xiatianrui1110 <xiatianrui1110@gmail.com>

xiatianrui1110 committed:
fix: Update evaluation scripts and results paths

Signed-off-by: xiatianrui1110 <xiatianrui1110@gmail.com>

phoevos committed:
wip: I'm going through changes

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
Merge branch 'main' into ci-config

phoevos committed:
fix: Add environment and indent

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Rename workflow

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fixup! fix: Add environment and indent

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Add job names

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Move cleanup to separate job and ensure it always runs

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
wip: Debug az/azd issues

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Update step names

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Make script executable?

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Make script executable; nope

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Provide necessary values for azd env refresh

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: azd auth login

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Skip deployment

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Make jobs run independently

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Make jobs run independently v2

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Setup azd for cleanup job

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Make cleanup into a step

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Move envvars to job level

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos commented:
I've followed the instructions found [here](https://learn.microsoft.com/en-us/azure/developer/github/connect-from-azure?tabs=azure-portal%2Clinux) to create a Microsoft Entra application for our CI deployments and generate federated credentials for it so that we don't use our own client ID. For now, I'm setting the branch to `ci-config` in the subject identifier, but once we merge this will only be `main`.

phoevos committed:
fix: Make workflow self-standing

Move application deployment to a step in the model evaluation workflow.
The previous approach failed to run the hooks defined in azure.yaml.
Most importantly, given that the postprovision scripts were not executed
prepdocs was never run and AI Search was never populated.

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Refresh env and avoid prompts

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Add missing envvars

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
feat: Make number of questions configurable

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Move client ID to a secret

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Standardise comment formatting (see body)

def generate_comment(text, line_length=60, comment_char='-'):
    padding = (line_length - len(text) - 2) // 2
    comment_line = f"{comment_char * padding} {text} {comment_char * padding}"
    if len(comment_line) < line_length:
        comment_line += comment_char
    return f"#{comment_line}"

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Fucking hell

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Simplify boolean input handling

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Refresh azd env

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos committed:
fix: Move env refresh after deployment

Refreshing the environment fails if the environment does not exist.
Therefore, if 'azd up' has run, this will be a noop.

Signed-off-by: Phoevos Kalemkeris <phoevos@protonmail.com>

phoevos commented:
@GrayNekoBean, PTAL at my changes. The commits won't help you much, cause I was testing the workflow and making changes on the fly, but it should be easy enough to tell what's altered by looking at the final diff. 

The main change introduced was making the workflow self-contained, i.e. handling the deployment and purging of the resources apart from the evaluation. Previously, I argued that we should be calling `azure-dev` to handle provisioning/deploying, but testing showed that's not feasible in our case! Running `provision` and `deploy` separately skips crucial steps of the pipeline defined in `azure.yaml`, most importantly postprovisioning, which entails running `prepdocs` to populate AI Search. This can be seen in this failed pipeline which reports that the [index was not found](https://github.com/ucl-contoso-chat/ucl-openai-search/actions/runs/10230218191/job/28304778126#step:11:76). The reason why that had never shown before was that the original version of the code relied on the synthetic data being committed to the repo (therefore skipping the generation step that actually makes use of AI Search). Of course, that approach was misguided; we are now running the generation script before proceeding with the evaluation steps.

The other major change involved cleaning up most of the environment/repo variables, since they only created a mess, making debugging impossible (this makes the workflow less configurable of course). Using the client ID retrieved from your local environment meant that I couldn't run the deployment for different environments. I believe I fixed that with the approach described [here](https://github.com/ucl-contoso-chat/ucl-openai-search/pull/11#issuecomment-2267114447). Was this PR ever actually tested e2e? Did it ever go through a fresh deployment? Did `azd` ever work (I had to add the separate login logic to allow using `azd`, e.g. when purging)? Note also how I'm using a script to generate a `.env` file based on the actual values generated by the deployment, thus not having to rely on values hardcoded in the repo environment. 

Apart from that, I reinstated a flag like the one you had before to skip deployments (arguably, it's much more convenient for testing, still doubt its practicality in any real use case) and I made the azd environment and number of questions for evaluation configurable. Finally, I made it so that purging can be triggered (when specified by the user) even in cases where other steps fail (crucial in order to avoid leaving stale resources behind when a deployment fails).

GrayNekoBean commented on .github/workflows/model-evaluation.yml:
```diff
@@ -0,0 +1,159 @@
+
+      - name: Purge deployment
+        if: ${{ always() && inputs.purge }}
+        run: azd down --force --purge
```
I think I forgot to fix this, this `azd down --force --purge` isn't going to work properly, because `azd down` doesn't accept these 2 arguments at the same time, which will still require manual operation. I went to the doc before and I think `azd down --force` should be the right thing to put here, but I haven't tested that yet. Could you please verify that?

phoevos replied:
I've been running it this way both locally and in the CI, seems to be working!
